server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -0,0 +1,366 @@
+     */
+    private Map<String, UploadedSegmentMetadata> segmentsUploadedToRemoteStore;
+
+    private static final Logger logger = LogManager.getLogger(RemoteSegmentStoreDirectory.class);
+
+    public RemoteSegmentStoreDirectory(RemoteDirectory remoteDataDirectory, RemoteDirectory remoteMetadataDirectory) throws IOException {
+        super(remoteDataDirectory);
+        this.remoteDataDirectory = remoteDataDirectory;
+        this.remoteMetadataDirectory = remoteMetadataDirectory;
+        init();
+    }
+
+    /**
+     * Initializes the cache which keeps track of all the segment files uploaded to the remote segment store.
+     * As this cache is specific to an instance of RemoteSegmentStoreDirectory, it is possible that cache becomes stale
+     * if another instance of RemoteSegmentStoreDirectory is used to upload/delete segment files.
+     * It is caller's responsibility to call init() again to ensure that cache is properly updated.
+     * @throws IOException if there were any failures in reading the metadata file
+     */
+    public void init() throws IOException {
+        this.metadataFileUniqueSuffix = UUIDs.base64UUID();
+        this.segmentsUploadedToRemoteStore = new ConcurrentHashMap<>(readLatestMetadataFile());
+    }
+
+    /**
+     * Read the latest metadata file to get the list of segments uploaded to the remote segment store.
+     * We upload a metadata file per refresh, but it is not unique per refresh. Refresh metadata file is unique for a given commit.
+     * The format of refresh metadata filename is: refresh_metadata__PrimaryTerm__Generation__UUID
+     * Refresh metadata files keep track of active segments for the shard at the time of refresh.
+     * In order to get the list of segment files uploaded to the remote segment store, we need to read the latest metadata file.
+     * Each metadata file contains a map where
+     *      Key is - Segment local filename and
+     *      Value is - local filename::uploaded filename::checksum
+     * @return Map of segment filename to uploaded filename with checksum
+     * @throws IOException if there were any failures in reading the metadata file
+     */
+    private Map<String, UploadedSegmentMetadata> readLatestMetadataFile() throws IOException {
+        Map<String, UploadedSegmentMetadata> segmentMetadataMap = new HashMap<>();
+
+        Collection<String> metadataFiles = remoteMetadataDirectory.listFilesByPrefix(MetadataFilenameUtils.METADATA_PREFIX);
+        Optional<String> latestMetadataFile = metadataFiles.stream().max(METADATA_FILENAME_COMPARATOR);
+
+        if (latestMetadataFile.isPresent()) {
+            logger.info("Reading latest Metadata file {}", latestMetadataFile.get());
+            segmentMetadataMap = readMetadataFile(latestMetadataFile.get());
+        } else {
+            logger.info("No metadata file found");
+        }
+
```

Comment:
what are the cases where no metadata is a valid case and when is it unexpected. Either add assertions or elevate warn logging as needed

Reply:
This can happen for the new shard. I will add more info to the message but we can still keep it as info. What do you think?

Reply:
can the caller at least assert?

Reply:
Yes, caller can still assert on the returned map that is returned by this method.